ocaml-lsp-server/src/code_actions/action_annotate.ml

```diff
@@ -0,0 +1,63 @@
+open Import
+
+let action_kind = "annotate"
+
+let check_context doc pos_start =
+  Document.with_pipeline doc (fun pipeline ->
+      let pos_start = Mpipeline.get_lexing_pos pipeline pos_start in
+      let typer = Mpipeline.typer_result pipeline in
+      let browse = Mbrowse.of_typedtree (Mtyper.get_typedtree typer) in
+      match Mbrowse.enclosing pos_start [ browse ] with
+      | (_, (Expression _ | Pattern _)) :: _ -> `Valid
+      | _ :: _
+      | [] ->
+        `Invalid)
+
+let get_source_text doc (loc : Loc.t) =
+  let open Option.O in
+  let source = Document.source doc in
+  let* start = Position.of_lexical_position loc.loc_start in
+  let+ end_ = Position.of_lexical_position loc.loc_end in
+  let (`Offset start) = Msource.get_offset source (Position.logical start) in
+  let (`Offset end_) = Msource.get_offset source (Position.logical end_) in
+  String.sub (Msource.text source) ~pos:start ~len:(end_ - start)
```

Comment:
You can use `String.drop` here.

Reply:
I don't think so, drop is only dropping the beginning of the string while I want to drop at the beginning and at the end?

Reply:
Nvm, you're right.